### PR TITLE
Refactor TensorMap constructors

### PR DESCRIPTION
**Pull request by lkdvos**
This PR is a refactor of the TensorMap constructors, which brings them more in line with the `Array` counterparts.

In particular, the default way of constructing uninitialized tensormap should now be:
```julia
TensorMap{E}(undef, codomain ← domain)
TensorMap{E}(undef, codomain, domain)
Tensor{E}(undef, space)
```
While they can also be constructed from actual data:
```julia
TensorMap(data, codomain ← domain)
TensorMap(data, codomain, domain)
```

For convenience, the following functions are also supported: `rand`, `randn`, `ones`, `zeros`
```julia
rand([T], codomain ← domain))
rand([T}, codomain, domain)
rand([T], codomain)
```

Finally, it is still possible to change the storagetype `A`, which does however require the fully specified type:
```julia
TensorMap{E,S,N1,N2,I,A,F1,F2}(undef, codomain, domain)
tensormaptype(S,N1,N2,EorA)(undef, codomain, domain) # equivalent
``` 

A similar approach is adopted for `isomorphism`, `id`, `unitary` and `isometry`:
```julia
isomorphism([TorA::Type=Float64,] codomain::TensorSpace, domain::TensorSpace) -> TensorMap
isomorphism([TorA::Type=Float64,] codomain ← domain) -> TensorMap
```
---

a couple questions:
- Should we try and make the default way of dealing with tensors via `HomSpace`, and reflect this as much as possible in the docstrings and documentation?
- There is a slight difference between constructing a tensormap with initialized blocks, where the blocks have uninitialized data, or a tensormap with uninitialized blocks. In principle, having this second option could be convenient, for example in the implementations of qr and svd, where LinearAlgebra already generates blocks.

Any feedback definitely appreciated.

### CODE REVIEW

File: src/tensors/linalg.jl [hunk 1]
```diff
@@ -45,106 +45,104 @@ function one!(t::AbstractTensorMap)
...
-                  cod::ProductSpace{S},
-                  dom::ProductSpace{S}) where {A<:DenseMatrix,S<:ElementarySpace}
+function isometry(::Type{A}, V::TensorMapSpace{S,N₁,N₂}) where {A<:MatOrNumber,S,N₁,N₂}
     InnerProductStyle(S) === EuclideanProduct() || throw_invalid_innerproduct(:isometry)
-    dom ≾ cod ||
-        throw(SpaceMismatch("codomain $cod and domain $dom do not allow for an isometric mapping"))
-    t = TensorMap(s -> A(undef, s), cod, dom)
-    for (c, b) in blocks(t)
+    domain(V) ≾ codomain(V) ||
+        throw(SpaceMismatch("codomain and domain are not isometric: $V"))
```
Comment (Jutho) [inline src/tensors/linalg.jl]: "isometric" means equivalent as metric spaces. For inner product spaces, this basically means that they are unitarily equivalent. Hence, I prefer the older terminology.

File: src/tensors/abstracttensor.jl [hunk 2]
```diff
@@ -226,6 +226,41 @@ Return the dimensions of the block of a tensor corresponding to a coupled sector
 Return an iterator over all splitting - fusion tree pairs of a tensor.
 """ fusiontrees(::AbstractTensorMap)
 
+# Similar
```
Comment (Jutho) [inline src/tensors/abstracttensor.jl:229]: Is there no doc string for `similar` of `AbstractTensorMap` yet?

**Comment by Jutho** [comment]
> * Should we try and make the default way of dealing with tensors via `HomSpace`, and reflect this as much as possible in the docstrings and documentation?

We can, but currently that is indeed not the case, also in the implementation. E.g. similar and others call `tensormaptype(...)(undef, codomain, domain)`. I guess that typing a comma is easier than typing a left pointing arrow with `\leftarrow`.

Looking at the docstring of `Matrix{T}`, it seems that only `Matrix{T}(undef, m, n)` is mentioned, even though `Matrix{T}(undef, (m, n))` definitely works.

> * There is a slight difference between constructing a tensormap with initialized blocks, where the blocks have uninitialized data, or a tensormap with uninitialized blocks. In principle, having this second option could be convenient, for example in the implementations of qr and svd, where LinearAlgebra already generates blocks.

Having uninitialized blocks seems like a very fragile state, that we do not want to expose to the user. Probably there are better solutions requiring less copies/allocations to fill the blocks in the case of factorisations.

**Comment by lkdvos** [comment]
> Looking at the docstring of `Matrix{T}`, it seems that only `Matrix{T}(undef, m, n)` is mentioned, even though `Matrix{T}(undef, (m, n))` definitely works.

Interestingly, when looking at `Array{T}`, the opposite is true, and only `Array{T}(undef, dims)` is mentioned, while the other cases also definitely work. In any case, it seems like `zeros` and similar mention both options so let's just keep them both as well.

---
I addressed the other comments, do you think this is ready to go?

**Review by Jutho — APPROVED** [review]

**Pull request merged**